Lib/test/test_thread.py

```diff
@@ -233,58 +233,60 @@ def task():
         with self.assertRaisesRegex(RuntimeError, "Cannot join current thread"):
             raise errors[0]
 
-    def test_detach_from_self(self):
-        errors = []
-        handles = []
-        start_joinable_thread_returned = thread.allocate_lock()
-        start_joinable_thread_returned.acquire()
-        thread_detached = thread.allocate_lock()
-        thread_detached.acquire()
+    def test_join_then_self_join(self):
+        # make sure we can't deadlock in the following scenario with
+        # threads t0 and t1:
+        #
+        # - t0 joins t1
+        # - t1 self joins
+        def make_lock():
+            lock = thread.allocate_lock()
+            lock.acquire()
+            return lock
+
+        error = None
+        self_joiner_handle = None
+        self_joiner_started = make_lock()
+        self_joiner_barrier = make_lock()
+        def self_joiner():
+            nonlocal error
+
+            self_joiner_started.release()
+            self_joiner_barrier.acquire()
 
-        def task():
-            start_joinable_thread_returned.acquire()
             try:
-                handles[0].detach()
+                self_joiner_handle.join()
             except Exception as e:
-                errors.append(e)
-            finally:
-                thread_detached.release()
+                error = e
+
+        joiner_started = make_lock()
+        def joiner():
+            joiner_started.release()
+            self_joiner_handle.join()
 
         with threading_helper.wait_threads_exit():
-            handle = thread.start_joinable_thread(task)
-            handles.append(handle)
-            start_joinable_thread_returned.release()
-            thread_detached.acquire()
-            with self.assertRaisesRegex(ValueError, "not joinable"):
-                handle.join()
+            self_joiner_handle = thread.start_joinable_thread(self_joiner)
+            # Wait for the self-joining thread to start
+            self_joiner_started.acquire()
 
-        assert len(errors) == 0
+            # Start the thread that joins the self-joiner
+            joiner_handle = thread.start_joinable_thread(joiner)
 
-    def test_detach_then_join(self):
-        lock = thread.allocate_lock()
-        lock.acquire()
+            # Wait for the joiner to start
+            joiner_started.acquire()
 
-        def task():
-            lock.acquire()
+            # Not great, but I don't think there's a deterministic way to make
```

Comment:
Unfortunately this is still racy, though probably less error prone than the previous approach. To be correct, we need to wait until the joining thread is blocked, not just called, before unblocking the self-joining thread. I don't think it's possible to do that (without adding a lot of extra complexity into the runtime).

Reply:
Ugh for sure.

There might be some creative^Wgross ways to figure that out via ctypes or testcapi? I dislike tests that sleep and hope for things to happen as it is best to assume they always wind up flaky and incur ongoing tweaking and random frustration.

Is it possible to introspect the thread at the OS level on one of our tier1 platforms to deterministically understand when it is actually blocked on the thing we want it blocked on?  (we'd be back to polling waiting for that to happen, but if it _is_ deterministic that would still be nice)

Reply:
> I dislike tests that sleep

Me too!

> Is it possible to introspect the thread at the OS level on one of our tier1 platforms to deterministically understand when it is actually blocked on the thing we want it blocked on?

This feels like it might end up being brittle in a different way. We'd end up relying on implementation details of `PyEvent` (i.e. putting the thread to sleep while blocked) and the OS.

Reply:
So unfortunately we can't even override `join` as it's read-only. I can do some digging into how difficult it would be to verify that the thread is actually blocked if you feel strongly, but I'm not sure it's worth the added complexity.

Reply:
makes sense, I like to _try_ to avoid timing related issues but sometimes how to do so is not clear cut.  we can revisit this test if it causes problems.